pallets/api/src/extension.rs

```diff
@@ -0,0 +1,121 @@
+use core::{fmt::Debug, marker::PhantomData};
+use frame_support::traits::Get;
+pub use pop_chain_extension::{Config, DecodingFailed, DispatchCall, ReadState, Readable};
+use pop_chain_extension::{
+	Converter, Decodes, Environment, Ext, LogTarget, Matches, Processor, Result, RetVal, State,
+};
+use sp_runtime::DispatchError;
+use sp_std::vec::Vec;
+
+/// The logging target for the chain extension.
+pub const LOG_TARGET: &str = "pop-api::extension";
+
+/// The chain extension used by the API.
+pub type Extension<Functions> = pop_chain_extension::Extension<Functions>;
+/// Decodes output by prepending bytes from ext_id() + func_id()
+pub type DecodesAs<Output, Error, Logger = ()> = Decodes<Output, Error, Prepender, Logger>;
+
+/// Prepends bytes from ext_id() + func_id() to prefix the encoded input bytes to determine the versioned output
+pub struct Prepender;
+impl Processor for Prepender {
+	/// The type of value to be processed.
+	type Value = Vec<u8>;
+
+	/// The log target.
+	const LOG_TARGET: &'static str = "pop-api::extension::processor";
+
+	/// Processes the provided value.
+	///
+	/// # Parameters
+	/// - `value` - The value to be processed.
+	/// - `env` - The current execution environment.
+	fn process<E: Ext, S: State>(
+		mut value: Self::Value,
+		env: &mut Environment<E, S>,
+	) -> Self::Value {
+		// TODO: revisit the ordering based on specced standard
+		// Resolve version, pallet and call index from environment
+		let version = env.func_id().to_le_bytes()[0];
+		let (pallet_index, call_index) = {
+			let bytes = env.ext_id().to_le_bytes();
+			(bytes[0], bytes[1])
+		};
+		// Prepend bytes
+		value.insert(0, version);
+		value.insert(1, pallet_index);
+		value.insert(2, call_index);
+		log::debug!(target: Self::LOG_TARGET, "prepender: version={version}, module={pallet_index}, call={call_index}");
+		value
+	}
+}
+
+/// Matches on the first byte of a function identifier only.
+pub struct IdentifiedByFirstByteOfFunctionId<T>(PhantomData<T>);
+impl<T: Get<u8>> Matches for IdentifiedByFirstByteOfFunctionId<T> {
+	fn matches<E: Ext, S: State>(env: &Environment<E, S>) -> bool {
+		let bytes = env.func_id().to_le_bytes();
+		bytes[1] == T::get()
```

Comment:
This is dependent on the redefinition of the encoding scheme used which was to be dealt with separately to this PR.

Agree that the current naming is not representative of what it is actually doing, was more a case of demonstrating a descriptive name for the type to make it easier to follow what is going on when it is used higher up the stack.

@Daan - can you prepare an internal draft writeup of how it will work for alignment please?  I can then amend here accordingly, with that encoding change happening separately to this PR.

Reply:
Have refactored these into functions to make it easier to update once encoding scheme finalised.